pynobo.py

```diff
@@ -352,86 +417,68 @@ def send_command(self, commands):
         """
         response = b''
         while response[-1:] != b'\r':
             try:
-                response += self.client.recv(bufsize)
-            except socket.timeout:
-                # handshake needs to be sent every < 30 sec, preferably every 14 seconds
-                if keep_alive:
-                    now = time.time()
-                    if now - self.last_handshake > 14:
-                        self.send_command([self.API.HANDSHAKE])
-                        self.last_handshake = now
-                else:
-                    break
+                response += await self.reader.read(bufsize)
             except ConnectionError as e:
                 self.logger.info('lost connection to hub (%s)', e)
-                self.socket_connected.clear()
+                await self.close()
                 break
 
         # Handle more than one response in one receive
-        response_list = str(response, 'utf-8').split('\r')
+        response_list = str(response.decode("utf-8")).split('\r')
         response = []
         for r in response_list:
             if r:
                 response.append(r.split(' '))
 
         return response
 
-    def socket_receive(self):
-        """The task running in daemon thread"""
-        while not self.socket_receive_exit_flag.is_set():
-            try:
-                if self.socket_connected.is_set():
-                    resp = self.get_response(True)
-                    for r in resp:
-                        self.logger.debug('received: %s', r)
-
-                        if r[0] == self.API.HANDSHAKE:
-                            pass # Handshake, no action needed
-
-                        elif r[0][0] == 'E':
-                            self.logger.error('error! what did you do? %s', r)
-                            #TODO: Raise something here?
-
-                        else:
-                            self.response_handler(r)
-
-                else:
-                    self.reconnect_hub()
-            except Exception as e:
-                # Ops, now we have real problems
-                self.logger.error("Unhandeled exception: %s", e, exc_info=1)
-                # Just disconnect (in stead of risking an infinite reconnect loop)
-                self.socket_receive_exit_flag.set()
-
-        self.logger.info('receive thread exited')
+    async def socket_receive(self):
+        try:
+            while True:
+                responses = await self.get_response()
+                for r in responses:
+                    self.logger.debug('received: %s', r)
+                    if r[0] == self.API.HANDSHAKE:
+                        pass # Handshake, no action needed
+                    elif r[0][0] == 'E':
+                        self.logger.error('error! what did you do? %s', r)
+                        # TODO: Raise something here?
+                    else:
+                        self.response_handler(r)
+                # TODO: Reconnect if necessary
```

Comment:
Can we check if `self.reader`/`self.writer` is None at this point, call `reconnect_hub` as a `self.loop.call_soon`? `reconnect_hub` would have to start by calling `stop`, so existing tasks do not run while reconnecting.

Reply:
I implemented reconnect the following way in `socket_receive`:

If `get_response` throws `ConnectionError` (e.g. connection reset by peer) or `IncompleteReadError` (connection closed), we will try to reconnect. I kept the logic to rediscover the hub before reconnecting, but moved the loop to inside the UDP client to avoid instantiating it every 10 seconds. I did change the logic to allow the IP to change (unless initially provided) in rediscover, but kept the full serial. I also added support for `discover=False` in reconnect, which seems to be missing in master.

We will never enter `socket_receive` unless we have successfully connected and retrieved initial state from the hub, so I think this acceptable reconnect logic.

I was able to force disconnects and test reconnect by increasing the keep alive interval to 35 seconds. I always get connection reset by peer and never a proper connection close from the hub, but both cases are handled.

For `discover=False` I tested various scenarios where I mimicked connection refused and  no route to host.

Reply:
After a lot of testing, I ended up to handle `OSError` instead of `ConncetionError`, and check for the following network related errors: `ECONNRESET`, `ECONNREFUSED`, `EHOSTUNREACH`, `EHOSTDOWN`,`ENETDOWN`, 
`ENETUNREACH`.